elf_load.c

```diff
@@ -356,26 +632,57 @@ static int elf_load_fdread(struct elf_prog *elf_prog, Elf *elf, int fd)
 		goto err_out;
 	}
 
+	if (unlikely(elf_getphnum(elf, &phnum) == 0)) {
+		elferr_err("%s: Failed to get number of program headers",
+			   elf_prog->name);
+		ret = -ENOEXEC;
+		goto err_free_img;
+	}
+
+	/* Get highest PHDR alignment */
+	for (phi = 0; phi < phnum; ++phi) {
+		if (gelf_getphdr(elf, phi, &phdr) != &phdr) {
+			elferr_warn("%s: Failed to get program header %"PRIu64"\n",
+				    elf_prog->name, (uint64_t) phi);
+			continue;
+		}
+		if (phdr.p_type != PT_LOAD)
+			continue;
+
+		if (pt_load_max_align < phdr.p_align)
+			pt_load_max_align = phdr.p_align;
+	}
+
+	UK_ASSERT(pt_load_max_align && PAGE_ALIGNED(pt_load_max_align));
+
+#if CONFIG_LIBPOSIX_MMAP
+	/* If we use mmap, let `elf_load_fdphdr` decide the vabase depending
+	 * on what mmap returns. For now we simply set it to what the ELF
+	 * header tells us. Ultimately, `elf_load_fdphdr` will update it for
+	 * us with the new address.
+	 */
+	elf_prog->entry = ehdr.e_entry;
+#else /* !CONFIG_LIBPOSIX_MMAP */
 	elf_prog->vabase = uk_memalign(elf_prog->a, __PAGE_SIZE,
```

Comment:
Should this be `pt_load_max_align`?

Reply:
What the... I thought I updated it... thanks :100: